### PR TITLE
[SFP-Refactor] Implement CMIS Low Power mode

### PR DESCRIPTION
**Pull request by aravindmani-1**
**Description**
Add LP mode support for CMIS devices.

**Motivation and Context**
Low Power mode can be set in CMIS devices by reading/writing in EEPROM.

**How Has This Been Tested?**
Tested in DellEMC Z9332f platform.

### CODE REVIEW

File: sonic_platform_base/sonic_xcvr/sfp_optoe_base.py [hunk 1]
```diff
@@ -128,6 +128,14 @@ def set_power_override(self, power_override, power_set):
     def get_eeprom_path(self):
         raise NotImplementedError
 
+    def get_lpmode(self):
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/sfp_optoe_base.py:131]: the return type should match with the return type mentioned in sfp_base.py for this api

File: sonic_platform_base/sonic_xcvr/sfp_optoe_base.py [hunk 2]
```diff
@@ -128,6 +128,14 @@ def set_power_override(self, power_override, power_set):
...
+        api = self.get_xcvr_api()
+        return api.get_lpmode() if api is not None else None
+
+    def set_lpmode(self, lpmode):
+        api = self.get_xcvr_api()
+        return api.set_lp_mode(lpmode) if api is not None else None
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/sfp_optoe_base.py:137]: the return type should match with the return type mentioned in sfp_base.py for this api

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 3]
```diff
@@ -833,15 +833,38 @@ def reset_module(self, reset = False):
...
+        if self.is_flat_memory() or not self.get_lpmode_support():
+            return False
+
+        lpmode_val = self.xcvr_eeprom.read(consts.MODULE_LEVEL_CONTROL)
+        if lpmode_val is not None:
+            if lpmode is True:
+                lpmode_val = lpmode_val | (1 << 4)
+            else:
+                lpmode_val = lpmode_val & ~(1 << 4)
+            return self.xcvr_eeprom.write(consts.MODULE_LEVEL_CONTROL, lpmode_val)
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:866]: please confirm module indeed entered low power via get_lpmode(). if not return false.

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 4]
```diff
@@ -833,15 +833,38 @@ def reset_module(self, reset = False):
         else:
             return True
 
-    def set_low_power(self, AssertLowPower):
+    def get_lpmode(self):
+        '''
+        Retrieves Low power module status
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:838]: comment 'Returns True if module in low power else returns False.'

**Review by prgeor — APPROVED** [review]

File: sonic_platform_base/sonic_xcvr/sfp_optoe_base.py (hunk 1 above)
Comment (aravindmani-1) [reply sonic_platform_base/sonic_xcvr/sfp_optoe_base.py:131]: the above change will help us to check whether API is implemented or not. Incase False is returned, it'd be difficult to catch where False is returned(whether the module is not in LP mode/ API is not defined).

**Review by prgeor — APPROVED** [review]

**Comment by prgeor** [comment]
@aravindmani-1 unit test is failing. please check

File: sonic_platform_base/sonic_xcvr/api/public/cmis.py [hunk 5]
```diff
@@ -843,6 +843,42 @@ def set_low_power(self, AssertLowPower):
...
+        if self.is_flat_memory() or not self.get_lpmode_support():
+            return False
+
+        lpmode = self.xcvr_eeprom.read(consts.TRANS_MODULE_STATUS_FIELD)
+        if lpmode is not None:
+            if lpmode.get('ModuleState') == 'ModuleLowPwr':
+                return True
+        return False
+
+    def set_lpmode(self, lpmode):
```
Comment (prgeor) [inline sonic_platform_base/sonic_xcvr/api/public/cmis.py:860]: please remove set_low_power() - duplicate of this API

File: tests/sonic_xcvr/test_cmis.py [hunk 6]
```diff
@@ -782,7 +782,7 @@ def test_get_supported_power_config(self, mock_response, expected):
     def test_reset_module(self):
         self.api.reset_module(True)
 
-    def test_set_low_power(self, ):
+    def test_set_low_power(self):
         self.api.set_low_power(True)
```
Comment (prgeor) [inline tests/sonic_xcvr/test_cmis.py:786]: the new api is 'set_lpmode'

**Review by prgeor — CHANGES_REQUESTED** [review]
test_set_laser_freq() is failing because, now we need to mock following APIs:-

```
     def test_set_laser_freq(self, input_param, mock_response):
+        self.api.is_flat_memory = MagicMock()
+        self.api.is_flat_memory.return_value = False
+        self.api.get_lpmode_support = MagicMock()
+        self.api.get_lpmode_support.return_value = False
+
```
Please modify the test case test_set_laser_freq() as above.

**Review by prgeor — APPROVED** [review]

**Pull request merged**